### PR TITLE
[DomTree] Fix root attachment in runDFS()

### PR DESCRIPTION
**Pull request by nikic**
Currently, runDFS() only sets the Parent of the DFS root if it is already in the NodeToInfo map. This doesn't matter if we're running DFS on the DT root, which doesn't have a parent anyway. However, when running on PDT roots, this means we end up keeping the parent at 0, rather than setting it to 1 for the virtual PDT root. Because the virtual root (nullptr) has the same value as the dummy value in NumToNode (nullptr) this happens to work out by accident right now.

I believe we should always be setting the parent in runDFS(), and adjust AttachToNum in doFullDFSWalk() to be 1 (the virtual root) for PDTs.

### CODE REVIEW

**Comment by dwblaikie** [comment]
Does this need test coverage? Or is the assert essentially adding test coverage? (eg: would adding the assert have fired at ToT without the other changes?)

**Comment by nikic** [comment]
> Does this need test coverage? Or is the assert essentially adding test coverage? (eg: would adding the assert have fired at ToT without the other changes?)

Yes, the assert would fire without the changes. The change is intended to be NFC in terms of externally observable behavior, it only fixes internal state that is incorrect in a way that ends up not mattering with the current implementation. I ran into it when implementing https://github.com/llvm/llvm-project/pull/73097.

**Comment by dwblaikie** [comment]
> Yes, the assert would fire without the changes. The change is intended to be NFC in terms of externally observable behavior, it only fixes internal state that is incorrect in a way that ends up not mattering with the current implementation. 

Ah, cool - thanks for the explanation!

File: llvm/include/llvm/Support/GenericDomTreeConstruction.h [hunk 1]
```diff
@@ -552,7 +553,7 @@ struct SemiNCAInfo {
 
     addVirtualRoot();
     unsigned Num = 1;
-    for (const NodePtr Root : DT.Roots) Num = runDFS(Root, Num, DC, 0);
+    for (const NodePtr Root : DT.Roots) Num = runDFS(Root, Num, DC, 1);
```
Comment (kuhar) [inline llvm/include/llvm/Support/GenericDomTreeConstruction.h:556]: Maybe we can pass `Num` here to make it clear that we consider is to be the last one used *and* the node for attachment?

**Review by kuhar — APPROVED** [review]
LGTM % making it clear how last num used and num for attachment are related

File: llvm/include/llvm/Support/GenericDomTreeConstruction.h (hunk 1 above)
Comment (nikic) [reply llvm/include/llvm/Support/GenericDomTreeConstruction.h:556]: I don't think that would be correct. Note that this code updates `Num` on each iteration. For the first root, both will be the same, but for the later one the attachment needs to stay at 1 while the last num increases.

Comment (kuhar) [reply llvm/include/llvm/Support/GenericDomTreeConstruction.h:556]: Ah yes yes.

**Review by kuhar — APPROVED** [review]

**Pull request merged**